src/imperative/cached_op.cc

```diff
@@ -212,82 +223,62 @@ nnvm::Graph Imperative::CachedOp::GetForwardGraph(
-    }
-  #else
-    if (check_match(g, match)) return g;
-  #endif
-
-  auto create_memory_plan = [recording](nnvm::Graph& g) {
-    const auto& idx = g.indexed_graph();
+  if (!match) {
+    g.attrs.erase("forward_mem_plan");
+    g.attrs.erase("full_mem_plan");
+  } else if (g.attrs.count(recording ? "full_mem_plan" : "forward_mem_plan")) {
+    return g;
+  }
 
-    StorageVector storage(idx.num_node_entries(), exec::kBadStorageID);
-    for (const auto i : idx.input_nodes()) storage[idx.entry_id(i, 0)] = exec::kExternalStorageID;
-    const auto& stypes = g.GetAttr<StorageTypeVector>("storage_type");
-    CHECK_EQ(stypes.size(), storage.size());
-    for (size_t i = 0; i < stypes.size(); i++) {
-      if (stypes[i] != kDefaultStorage)
-        storage[i] = exec::kDynamicStorageID;
-    }
 
-    auto mem_plan = PlanMemory(
-        &g, std::move(storage), g.GetAttr<std::vector<uint32_t> >(
-            recording ? "full_ref_count" : "forward_ref_count"));
-    g.attrs[recording ? "full_mem_plan" : "forward_mem_plan"] =
-        std::make_shared<dmlc::any>(std::move(mem_plan));
-  };
 
 #if MXNET_USE_NGRAPH == 1
-  if (!recording) {
     ngraph_bridge::BindArgBase bind(num_inputs());
     ngraph_bridge::Compiler compiler(
-        fwd_graph_, inputs[0]->ctx(), cached_shape_inputs, cached_dtype_inputs,
+        fwd_graph_orig_, inputs[0]->ctx(), inputs_, cached_shape_inputs, cached_dtype_inputs,
         cached_storage_type_inputs);
 
-    ngraph_fwd_graph_ = compiler.Compile();
-    CheckAndInferShape(&ngraph_fwd_graph_, std::move(cached_shape_inputs),
+    g = compiler.Compile();
+    CheckAndInferShape(&g, std::move(cached_shape_inputs),
                        true);
-    CheckAndInferType(&ngraph_fwd_graph_, std::move(cached_dtype_inputs), true);
+    CheckAndInferType(&g, std::move(cached_dtype_inputs), true);
     exec::DevMaskVector cached_dev_mask(
-        ngraph_fwd_graph_.indexed_graph().num_nodes(),
+        g.indexed_graph().num_nodes(),
         inputs[0]->ctx().dev_mask());
-    CheckAndInferStorageType(&ngraph_fwd_graph_, std::move(cached_dev_mask),
+    CheckAndInferStorageType(&g, std::move(cached_dev_mask),
                              std::move(cached_storage_type_inputs), true);
 
-    const auto& idx = ngraph_fwd_graph_.indexed_graph();
-
+    const auto& idx = g.indexed_graph();
     std::vector<uint32_t> ref_count(idx.num_node_entries(), 0);
     for (const auto& i : idx.input_nodes()) ++ref_count[idx.entry_id(i, 0)];
     for (const auto& i : idx.outputs()) ++ref_count[idx.entry_id(i)];
     for (size_t i = 0; i < idx.num_nodes(); ++i) {
       for (const auto& j : idx[i].inputs) ++ref_count[idx.entry_id(j)];
     }
-
-    ngraph_fwd_graph_.attrs["forward_ref_count"] =
+    g.attrs["forward_ref_count"] =
         std::make_shared<dmlc::any>(ref_count);
-    create_memory_plan(ngraph_fwd_graph_);
-    return ngraph_fwd_graph_;
+
+    full_graph_ = GetFullGraph(g, compiler.GetInputs());
```

Comment:
You do this so you can run the Gradient after fusing the nGraph operations, correct?